### ISSUE TITLE
[FR] Plate subsampling without shuffling randomly the indexes

### ISSUE THREAD

**Issue by LysSanzMoreta** [pyro-ppl/pyro#2920]
Hi!

I am working with a type of data that would benefit of subsampling at the plate but without shuffling the data indexes. Is it possible to add a flag to the plate that says something like _shuffle=False_ ?

I did a small attempt to change the  [_Subsample class](https://github.com/pyro-ppl/pyro/blob/b7ba915b851d51c61d805da741bf7a74fcf9319d/pyro/poutine/subsample_messenger.py#L14-L64) but the change needs to happen elsewhere upstream and I soon lost track of the trace. Unless I am missing something....

It wil possibly be a quick fix for you and might take me hours (that I don't have right now unfortunately), that's why I pleaaaaase ask you to  consider it ,

Thanks in advance and highly appreciating all your efforts :)

**Comment by LysSanzMoreta** [comment]
@fritzo @martinjankowiak 

Hi! Sorry for late reply, back from vaccine sickness.
 
@martinjankowiak  I have a "special" model where the first part requires the full input data for some computation and  in the second part it is split. If I use mini-batching, the first part is computed n times (where n = number of mini-batches) and the model becomes quite slow. You could be right and there is another way around it but right now  I cannot use mini-batch ( I tried) and that's why I though about plating without shuffling  the indexes.

@fritzo I like subsample_size naming, but that's just my humble opinion :)

Thanks again!!!

**Comment by fritzo** [comment]
@LysSanzMoreta You should still be able to accomplish this passing `subsample_size` or `subsample`. Could you paste a concrete example model?

**Comment by LysSanzMoreta** [comment]
Hi!

I can give a guide of the full model in private :)

I have something like this;

``` python
def model(data):

      x = function_1 (data)
     
     with pyro.plate("data_plate", size = len(data), subsample_size = 20) as indx:

                    # Here indx have random order , I would like them to just be [0,1,2,3,4...,10], [....],[....,len(data)]]
                    
                   y = function_2(data[indx], info[indx])
                   aa = pyro.sample("obs",dist.Categorical(y), obs = data[indx])
```

Does that help?

Thanks! :)

**Comment by fritzo** [comment]
You can manually batch your model by passing in both `data` and the batch indices `indx`:
```py
def model(data, indx):
    x = function_1(data)
    with pyro.plate("data_plate", size=len(data), subsample=indx):
        y = function_2(data[indx], info[indx])
        aa = pyro.sample("obs",dist.Categorical(y), obs=data[indx])

guide = AutoNormal(model)  # or whatever
svi = Trace_ELBO(model, guide, ..., Trace_ELBO())
batch_size = 20
for epoch in range(10):
    for batch_start in range(0, len(data), batch_size):
        indx = torch.arange(batch_start, min(batch_start + batch_size, len(data)))
        svi.step(data, indx)
```
You could also avoid some indexing math by using [PyTorch dataloaders](https://pytorch.org/docs/stable/data.html#torch.utils.data.BatchSampler)
```py
from torch.utils.data import BatchSampler, SequentialSampler
for epoch in range(10):
    for indx in BatchSampler(SequentialSampler(range(len(data)), batch_size=batch_size)
        indx = torch.tensor(indx)
        svi.step(data, indx)
```

**Comment by LysSanzMoreta** [comment]
@fritzo  Ohh, I missed that ... I did not see that subsample could take in the indexes. I will try it out then. Thanks!
